scripts/ci/pre_commit/pre_commit_new_session_in_provide_session.py

```diff
@@ -0,0 +1,123 @@
+
+    Old code written before the introduction of NEW_SESSION (and even some new
+    if the contributor wasn't made aware of the addition) generally uses None
+    as the default value, so we add that to the check as well.
+    """
+    if isinstance(value, ast.Constant) and value.value is None:
+        return _SessionDefault.none
+    if isinstance(value, ast.Name) and value.id == "NEW_SESSION":
+        return _SessionDefault.new_session
+    # It's possible to do FOO = NEW_SESSION and reference FOO to work around
+    # this check, but let's rely on reviewers to catch this kind of shenanigans.
+    return None
+
+
+_ALLOWED_DECORATOR_NAMES = ("overload", "provide_session", "abstractmethod")
+
+
+def _is_decorated_correctly(nodes: list[ast.expr]) -> bool:
+    """Whether expected decorators are provided.
+
+    Three decorators would allow NEW_SESSION usages:
+
+    * ``@provide_session``: The canonical case.
+    * ``@overload``: A typing overload and not something to actually execute.
+    * ``@abstractmethod``: This will be overridden in a subclass anyway.
+    """
+    # This only accepts those decorators literally. Should be enough?
+    return any(isinstance(node, ast.Name) and node.id in _ALLOWED_DECORATOR_NAMES for node in nodes)
+
+
+def _annotation_has_none(value: ast.expr | None) -> bool:
+    if value is None:
+        return False
+    if isinstance(value, ast.Constant) and value.value is None:
+        return True
+    if isinstance(value, ast.BinOp) and isinstance(value.op, ast.BitOr):  # Type union.
+        return _annotation_has_none(value.left) or _annotation_has_none(value.right)
+    return False
+
+
+def _iter_incorrect_new_session_usages(path: pathlib.Path) -> typing.Iterator[ast.FunctionDef]:
+    """Check NEW_SESSION usages outside functions decorated with provide_session."""
+    for node in ast.walk(ast.parse(path.read_text("utf-8"), str(path))):
+        if not isinstance(node, ast.FunctionDef):
+            continue
+        session = _get_session_arg_and_default(node.args)
+        if session is None or session.default is None:
+            continue  # No session argument or the argument has no default, we're good.
+        if _is_decorated_correctly(node.decorator_list):
+            continue  # Has @provide_session so the default is expected.
+        default_kind = _is_new_session_or_none(session.default)
+        if default_kind is None:
+            continue  # Default value is not NEW_SESSION or None.
+        if default_kind == _SessionDefault.none and _annotation_has_none(session.argument.annotation):
+            continue  # None is OK if the argument is explicitly typed as None.
+        yield node
+
+
+def main(argv: list[str]) -> int:
+    paths = (pathlib.Path(filename) for filename in argv[1:])
+    errors = [(path, error) for path in paths for error in _iter_incorrect_new_session_usages(path)]
+    if errors:
+        print("Incorrect NEW_SESSION usages:", end="\n\n")
```

Comment:
I think we should explain the usages in detail - basically the comment you provided above should made it into message explaining what is wrong. The goal should be that those who made mistake and their static check fails should see what they should do to fix it. (rich-colorful if possible - I usually use Yellow to display this message in UI. Similarly the "Incorrect NEW_SESSION" shoud be in"[red]" - this way it is immediately visible in the CI output what is an error and what is the action to be executed.

This has the nice effect that you will not have to have comments any more.

Reply:
+1

Reply:
We should add a section in https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#coding-style-and-best-practices and print a link here

Reply:
Actually we already do have a section on this! I’ll just link it here.

https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#database-session-handling